pkg/debug/shipper/shipper.go

```diff
@@ -0,0 +1,197 @@
+package shipper
+
+import (
+	"bytes"
+	"context"
+	"crypto/ecdsa"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/kolide/krypto/pkg/echelper"
+	"github.com/kolide/launcher/ee/consoleuser"
+	"github.com/kolide/launcher/ee/control"
+	"github.com/kolide/launcher/pkg/agent"
+	"github.com/kolide/launcher/pkg/agent/types"
+	"github.com/kolide/launcher/pkg/launcher"
+)
+
+type shipperOption func(*shipper)
+
+// WithUploadURL causes the shipper to upload to the given url instead of requesting a url to upload to
+func WithUploadURL(url string) shipperOption {
+	return func(s *shipper) {
+		s.uploadURL = url
+	}
+}
+
+type shipper struct {
+	bytes.Buffer
```

Comment:
I think we can go with this. But you shouldn't need to buffer. 

Ought be able to create the initial http stuff in `New` and use an `io.Pipe` to connect the reader with a `Writer` here.

That would move a lot of `ship` into `New` a little into `Write` and a little into `Close`

Reply:
@directionless , take a look now please, was this what you had in mind? It's pretty cool, thank you for the suggestion I learned something today.

Reply:
okay, I think this is cleaned up pretty good